specification/transport_layer/can.tex

```diff
@@ -23,235 +23,252 @@ \subsection{CAN ID structure}
 summarize the purpose of the fields and their permitted values
 for message transfers, anonymous message transfers, and service transfers, respectively.
 
-
 % Please do not remove the hard placement specifier [H], it is needed to keep elements ordered.
 \begin{figure}[H]
-\centering
-\resizebox{\textwidth}{!}{
-\begin{tabular}{|l|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|c|}  \hline
-\multirow{3}{*}{\textbf{Anonymous}} &
-\multicolumn{3}{c|}{\multirow{2}{*}{Priority}} &
-\multicolumn{5}{c|}{\multirow{2}{*}{Reserved}} &
-\multicolumn{3}{c|}{\multirow{2}{*}{Subject ID}} &
-\multicolumn{10}{c|}{\multirow{2}{*}{Discriminator}} &
-\multicolumn{8}{c|}{Protocol Version} \\
-\cline{23-29}
-
- &
-\multicolumn{3}{c|}{} & \multicolumn{5}{c|}{} & \multicolumn{3}{c|}{} & \multicolumn{10}{c|}{} &
-\multicolumn{7}{c|}{Anonymous Node ID} &  \\
-
- &
-\multicolumn{3}{c|}{[0, 7]} &
-\multicolumn{5}{c|}{0x00} &
-\multicolumn{3}{c|}{[0, 7]} &
-\multicolumn{10}{c|}{[0, 1023]} &
-\multicolumn{7}{c|}{0x00} & 1 \\ \hline
-
-\textbf{CAN ID Bit} &
-28 & 27 & 26 & 25 & 24 & 23 & 22 & 21 & 20 & 19 & 18 & 17 & 16 & 15 &
-14 & 13 & 12 & 11 & 10 &  9 &  8 & 7 &  6 &  5 &  4 &  3 &  2 &  1 &  0 \\
-\hline
-
-\multicolumn{30}{c}{} \\ \hline %% Table seperator
-
-\multirow{3}{*}{\textbf{Message}} &
-\multicolumn{4}{c|}{service/!message} &
-\multicolumn{5}{c|}{Reserved} &
-\multicolumn{12}{c|}{\multirow{2}{*}{Subject ID}} &
-\multicolumn{8}{c|}{Protocol Version} \\
-\cline{2-4} \cline{7-10} \cline{23-29}
-
- &
-\multicolumn{3}{c|}{Priority}
- &
- &
- &
-\multicolumn{16}{c|}{} &
-\multicolumn{7}{c|}{Source Node ID} &
-\\
-
- &
-\multicolumn{3}{c|}{[0, 7]} &
-0 &
-0 &
-\multicolumn{4}{c}{} &
-\multicolumn{12}{c|}{[0, 65535]} &
-\multicolumn{7}{c|}{[1, 127]} & 1 \\
-\hline
-
-\textbf{CAN ID Bit} &
-28 & 27 & 26 & 25 & 24 & 23 & 22 & 21 & 20 & 19 & 18 & 17 & 16 & 15 &
-14 & 13 & 12 & 11 & 10 &  9 &  8 &  7 &  6 &  5 &  4 &  3 &  2 &  1 &  0 \\
-\hline
-
-\multicolumn{30}{c}{} \\ \hline %% Table seperator
-
-\multirow{3}{*}{\textbf{Service}} &
-\multicolumn{4}{c|}{service/!message} &
-\multicolumn{5}{c|}{request/!response} &
-\multicolumn{5}{c|}{} &
-\multicolumn{7}{c|}{\multirow{2}{*}{Destination Node ID}} &
-\multicolumn{8}{c|}{Protocol Version} \\
-\cline{2-4} \cline{7-10} \cline{23-29}
-
- &
-\multicolumn{3}{c|}{Priority} &
-\cellcolor{red} &
- &
-\multicolumn{9}{c|}{Service ID} &
-\multicolumn{7}{c|}{} &
-\multicolumn{7}{c|}{Source Node ID} &  \\
- &
-\multicolumn{3}{c|}{[0, 7]} &
-1 &
-0|1 &
-\multicolumn{9}{c|}{[0, 511]} &
-\multicolumn{7}{c|}{[1, 127]} &
-\multicolumn{7}{c|}{[1, 127]} & 1 \\
-\hline
-
-\textbf{CAN ID Bit} &
-28 & 27 & 26 & 25 & 24 & 23 & 22 & 21 & 20 & 19 & 18 & 17 & 16 & 15 &
-14 & 13 & 12 & 11 & 10 &  9 &  8 &  7 &  6 &  5 &  4 &  3 &  2 &  1 &  0 \\
-\hline
-
-\end{tabular}
-}
-
-\caption{CAN ID structure}\label{fig:can_id_structure}
+    \centering
```

Comment:
I'm super-nervous that, should this protocol see widespread use, the 128 node limit will vex us. I'll defer to your judgment here but my main concern arises from using UAVCAN on linux. SocketCAN creates a virtual CAN network which makes it easy to suddenly have hundreds of nodes if the system is designed like a ROS system where there are a multitude of small, interconnected functions. Using such an architecture and for a complex system it'd be easy to run out of node identifiers. Are we thinking that such a system is ill-advised or do we think that a NAT-like scheme might be needed to support this type of use case (i.e. where each linux network was isolated somehow by router nodes that could forward CAN messages across physical busses?

Reply:
Seems reasonable.